content/blog/using-structured-outputs-in-vllm.md

```diff
@@ -0,0 +1,179 @@
+
+Here’s how each works, along with example outputs:
+
+### **1. Guided Choice**
+
+Simplest form of structured output, ensuring the response is one of a set of predefined options.
+
+```python
+from openai import OpenAI
+
+client = OpenAI(base_url="http://localhost:8000/v1", api_key="-")
+
+completion = client.chat.completions.create(
+    model="Qwen/Qwen2.5-3B-Instruct",
+    messages=[
+        {"role": "user", "content": "Classify this sentiment: vLLM is wonderful!"}
+    ],
+    extra_body={"guided_choice": ["positive", "negative"]},
+)
+print(completion.choices[0].message.content)
+```
+
+**Example Output:**
+
+```
+positive
+```
+
+### **2. Guided Regex**
+
+Constrains output to match a regex pattern, useful for formats like email addresses.
+
+```python
+completion = client.chat.completions.create(
+    model="Qwen/Qwen2.5-3B-Instruct",
+    messages=[
+        {
+            "role": "user",
+            "content": "Generate an example email address for Alan Turing at Enigma. End in .com.",
+        }
+    ],
+    extra_body={"guided_regex": r"\w+@\w+\.com\n", "stop": ["\n"]},
+)
+print(completion.choices[0].message.content)
+```
+
+**Example Output:**
+
+```
+alan.turing@enigma.com
+```
+
+### **3. Guided JSON**
+
+Enforces a valid JSON format based on a schema, simplifying integration with other systems.
+
+```python
+from pydantic import BaseModel
+from enum import Enum
+
+class CarType(str, Enum):
+    sedan = "sedan"
+    suv = "SUV"
+    truck = "Truck"
+    coupe = "Coupe"
+
+class CarDescription(BaseModel):
+    brand: str
+    model: str
+    car_type: CarType
+
+json_schema = CarDescription.model_json_schema()
+
+completion = client.chat.completions.create(
+    model="Qwen/Qwen2.5-3B-Instruct",
+    messages=[
+        {"role": "user", "content": "Generate a JSON for the most iconic car from the 90s."}
+    ],
+    extra_body={"guided_json": json_schema},
+)
+print(completion.choices[0].message.content)
+```
+
+**Example Output:**
+
+```json
+{
+  "brand": "Toyota",
+  "model": "Supra",
+  "car_type": "coupe"
+}
+```
+
+### **4. Guided Grammar**
+
+Uses an EBNF grammar to define complex output structures, such as SQL queries.
+
+```python
+completion = client.chat.completions.create(
+    model="Qwen/Qwen2.5-3B-Instruct",
+    messages=[
+        {"role": "user", "content": "Generate a SQL query to find all users older than 30."}
+    ],
+    extra_body={
+        "guided_grammar": """
+        query ::= "SELECT" fields "FROM users WHERE" condition;
+        fields ::= "name, age" | "*";
+        condition ::= "age >" number;
+        number ::= [0-9]+;
+        """
+    },
+)
+print(completion.choices[0].message.content)
+```
+
+**Example Output:**
+
+```sql
+SELECT * FROM users WHERE age > 30;
+```
+
+## **Next Steps**
+
+To start integrating structured outputs into your projects:
+
+1. **Explore the Documentation:** Check out the official documentation for more examples and detailed explanations.
```

Comment:
Sentence case:
**Explore the documentation:**

Also, I suggest to add a hotlink to the "Explore the documentation:" Check out the [official documentation](HTTP URL to the doc) for more examples and detailed explanations.